_posts/2020-08-05-hive-acid-original-files.md

```diff
@@ -0,0 +1,182 @@
+original files, the original transaction ID is always 0.
+
+### Bucket ID
+
+Bucket ID is retrieved from the original file name. For the original file
+`0000ABC_DEF` or `0000ABC_DEF_copy_G`, the bucket ID will be `ABC`.
+
+### Row ID
+
+To calculate the row ID, the total row count of all the original files, which
+come before the current one in lexicographical order, is calculated.
+Then, the row ID is equal to the sum of the value calculated and local row ID in
+the current original file.
+
+Here is an example to calculate the global Row ID of the 3rd row of an original
+File `000000_0_copy_2`.
+~~~~
+000000_0            -> 	X1 Rows (returned by ORC footer field numberOfRows)
+
+000000_0_copy_1     -> 	X2 Rows (returned by ORC footer field numberOfRows)
+
+000000_0_copy_2     ->	[ Row 0 ]
+                        [ Row 1 ]
+                        [ Row 2 ]   <- Local Row ID (returned filePosition in OrcRecordReader) = 2, Global Row ID = (X1+X2+2)
+                        [ Row 3 ]
+
+000000_0_copy_3     ->  X4 Rows
+~~~~
+
+>**Note:** As we see, additional computations are required to generate row IDs
+>while reading original files, therefore, read is slower than ACID format files
+>in the transactional
+table.
+
+Once Presto has the 3 ACID columns for a row, it can check for update/delete on
+it. Delete deltas, written by Hive for Original files, have row IDs generated by
+following the same
+strategy as discussed above, hence, the same logic of filtering out deleted rows
+as discussed in [Hive ACID and transactional tables' support in Presto
+] {% post_url 2020-06-01-hive-acid %} works with the original files too.
+
+# Changes in Presto to support reading original files
+
+Presto split generation logic and ORC reader is modified to add read support
+for original files. Following are the changes done at coordinator and worker
+level:
+
+## Split generation
+
+We use a new class named `AcidInfo` to store `OriginalFiles
+`, `DeleteDeltaFiles` for HiveSplit. `BackgroundSplitLoader.loadPartitions
+` is called in an executor to create splits for each partition. In addition
+to the steps mentioned in blog [Hive ACID and transactional tables' support in
+Presto] {% post_url 2020-06-01-hive-acid %}, Presto does the following:
+
+1. Original files, ACID subdirectories (`base`, `delta`, `delete_delta`) are
+figured out by listing the partition location by Hive AcidUtils Helper class.
+2. Registry for delete deltas `DeleteDeltaInfo` is created which has minimal
+information through which `delete_delta` path can be constructed by the workers.
+3. Registry for original files `OriginalFileInfo` is created which has
+information like file name, size and bucket ID.
+4. `AcidInfo.Builder` keeps a map
+`AcidInfo.Builder.bucketIdToOriginalFileInfoMap` of bucket ID to the list of
+original files belonging to the same bucket.
+5. Hive splits are created for each original file, base and delta directories.
```

Comment:
Maybe break this up ... three nested lists seems a bit much

Reply:
I have added them in separate lines to reduce the nesting.